lib/styleToFilters.js

```diff
@@ -42,13 +45,98 @@ function styleToFilters(style) {
         }
+
+        // Collect the used property
+        layers[layerName].properties = layers[layerName].properties || [];
+        ['paint', 'layout'].forEach(item => {
+          let itemObject = style.layers[i][item];
+          itemObject && getPropertyLoop(itemObject, layers[layerName].properties);
+        });
       }
     }
   }
 
+  // remove duplicate propertys and fix choose all the propertys in layers[i].properties
+  Object.keys(layers).forEach(layerId => {
+    let properties = layers[layerId].properties;
+    if (properties.indexOf(true) !== -1) {
+      layers[layerId].properties = true;
+    } else {
+      let unique = {};
+      properties.forEach(function(i) {
+        if (!unique[i]) {
+          unique[i] = true;
+        }
+      });
+      layers[layerId].properties = Object.keys(unique);
+    }
+  })
+
   return layers;
 }
 
+
+function getPropertyLoop(propertyObj, properties) {
+  Object.keys(propertyObj).forEach(key => {
+    let value = propertyObj[key];
+    // TODO we still have outher situation:
+    // - special properties: `mapbox_clip_start`, `mapbox_clip_end`
+    if (typeof value === 'string') {
+      // if the value is string try to get property name from `xx{PropertyName}xx` like.
+      // the /{[^}]+}/ig return all the value like {xxx}
+      // eg 'a{hello}badfa' => ['{hello}']
+      // eg 'a{hello}ba{world}dfa' => ['{hello}','{world}']
+      let preProperties = value.match(/{[^}]+}/ig);
+      preProperties && preProperties.forEach(item => {
+        properties.push(item.slice(1, -1));
+      });
+    } else if (typeof value === 'object' && typeof value.property === 'string') {
+      // - legacy functions with `property`
+      properties.push(value.property);
+    } else {
+      // test isExpression from sytleSpec
+      if (styleSpec.expression.isExpression(value)) {
+        // TODO: now we implement this by ourself in vtshavem, we need to talk with ‘style spec’ member to see if there have a official method to get used property, to make this can be synchronized with the expression update.
+        getPropertyFromExpression(value, properties);
+      } else {
+        // otherwise continual loop;
+        getPropertyLoop(value, properties);
+      }
+    }
+  })
+}
+
+
+function getPropertyFromExpression(exp, properties) {
+  // now we care about the expression like:
+  // ["get", string] not ["get", string, Object],
+  // ["has", string] not ["has", string, Object],
+  // ["properties"],
+  // ["feature-state", string]
+  if (exp instanceof Array) {
+    switch (exp[0]) {
+      case 'get':
+      case 'has':
+        if (typeof exp[1] === 'string' && !(exp[2] && typeof exp[2] === 'object')) {
+          properties.push(exp[1]);
+        }
+        break;
+      case 'feature-state':
+        properties.push(exp[1]);
+        break;
+      case 'properties':
+        properties.push(true);
+        break;
+    }
+
+    exp.forEach(sub => {
+      if (sub instanceof Array) {
```

Comment:
We check for `instanceof Array` inside of `getPropertyFromExpression`. Maybe only do it one place?

Reply:
hummm, thinking....
`getPropertyFromExpression` designed to receive Array, even in the 1st line of the function we have `if (exp instanceof Array) {`, 
but I think we should prevent pass an NON-ARRAY param to this founction, so inside the function while before we start  recursive(call the `getPropertyFromExpression`) I write this `.
If (sub instanceof Array) {`, if we remove this, the whole program also work, but that way we may break `the function designed to receive Array, we should prevent pass an NON-ARRAY param to this founction,`. 
It seems becomes a philosophical question again. LOL.

Reply:
Okay, if philosophical only, just make the decision yourself - I trust your judgement.

Reply:
Hmm, I choose the `if`.
`&&` it looks like a "hack trick", and the `if` is more common and better for others to read.

Reply:
oops, I got wrong content, I will keep the if check before recall the main function.